README.md
# MMM-ImagesPhotos

This is a module for the [MagicMirror](https://github.com/MichMich/MagicMirror). It will show photos from a directory.

This module reads the images from the _uploads_ directory inside the module.
**Directory**: _MagicMirror/modules/MMM-ImagesPhotos/uploads_

## Installation

1. Clone this repository inside your MagicMirror's `modules` folder

```bash
cd ~/MagicMirror/modules
git clone https://github.com/roramirez/MMM-ImagesPhotos
npm install
```

## How it looks

![Demo](.github/animate.gif)

## Config

The entry in `config.js` can include the following options:

<!-- prettier-ignore-start -->
| Option             | Description
|--------------------|-----------
| `opacity`          | The opacity of the image.<br><br>**Type:** `double`<br>Default 0.9
| `animationSpeed`   | How long the fade out and fade in of photos should take.<br><br>**Type:** `int`<br>Default 500
| `updateInterval`   | How long before loading a new image.<br><br>**Type:** `int`(milliseconds) <br>Default 5000 milliseconds
| `getInterval`      | Interval value to get new images from directory.<br><br>**Type:** `int`(milliseconds) <br>Default 60000 milliseconds
| `maxWidth`         | Value for maximum width. Optional, possible values: absolute (e.g. "700px") or relative ("50%") <br> Default 100%
| `maxHeight`        | Value for maximum height. Optional, possible values: absolute (e.g. "400px") or relative ("70%") <br> Default 100%
| `backgroundColor`  | Value for color used to fill around the image if not fullscreen, default 'black', can be #xxyyzz, like #808080 (grey),<br> if fill is true, the backgroundColor setting is ignored
| `fill`             | true or false, default false. instead of color use a blurred copy of the image to fill around the image
| `blur`             |  the size of the pixel blur of the background fill, default 8
<!-- prettier-ignore-end -->

Here is an example of an entry in `config.js`

```js
{
 module: "MMM-ImagesPhotos",
 position: "middle_center",
 config: {
  opacity: 0.9,
  animationSpeed: 500,
  updateInterval: 5000,
  backgroundColor: 'grey',  // not used if fill is true
  fill: false,   // fill around image with blurred copy of image
  blur: 10,   // only used if fill is true
 }
},
```


CHANGELOG.md
# Changelog

## [0.1.1] - 2017-03-13

- Update README documentation

## [0.1.0] - 2017-02-05

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-ImagesPhotos
 *
 * By Rodrigo Ramìrez Norambuena https://rodrigoramirez.com
 * MIT Licensed.
 */

const express = require("express");
const NodeHelper = require("node_helper");
const path = require("path");
const fs = require("fs");
const mime = require("mime-types");
const Log = require("logger");

let directoryImages;

module.exports = NodeHelper.create({
  // Override start method.
  start() {
    Log.log(`Starting node helper for: ${this.name}`);
  },

  setConfig() {
    this.path_images = path.resolve(
      `${global.root_path}/modules/MMM-ImagesPhotos/uploads${this.config.path}`
    );
    if (this.config.debug) {
      Log.log(`path for : ${this.name}= ${this.path_images}`);
    }
  },

  // Override socketNotificationReceived method.
  socketNotificationReceived(notification, payload) {
    if (notification === "CONFIG") {
      this.config = payload;
      this.setConfig();
      this.extraRoutes();
      this.sendSocketNotification("READY");
    }
  },

  // create routes for module manager.
  // recive request and send response
  extraRoutes() {
    const self = this;

    this.expressApp.get("/MMM-ImagesPhotos/photos", function (req, res) {
      self.getPhotosImages(req, res);
    });

    this.expressApp.use(
      "/MMM-ImagesPhotos/photo",
      express.static(self.path_images)
    );
  },

  // return photos-images by response in JSON format.
  getPhotosImages(req, res) {
    directoryImages = this.path_images;
    const imagesPhotos = this.getImages(this.getFiles(directoryImages)).map(
      function (img) {
        if (this.config.debug) {
          Log.log(`have image=${img}`);
        }
        Log.log(`/MMM-ImagesPhotos/photo/${img}`);
        return { url: `/MMM-ImagesPhotos/photo/${img}` };
      }
    );
    res.send(imagesPhotos);
  },

  // return array with only images
  getImages(files) {
    const images = [];
    const enabledTypes = ["image/jpeg", "image/png", "image/gif"];
    for (const idx in files) {
      const type = mime.lookup(files[idx]);
      if (enabledTypes.indexOf(type) >= 0 && type !== false) {
        images.push(files[idx]);
      }
    }

    return images;
  },

  getFiles(imagePath) {
    let files = [];
    try {
      files = fs.readdirSync(imagePath).filter(function (file) {
        if (!fs.statSync(`${imagePath}/${file}`).isDirectory()) {
          return file;
        }
      });
    } catch (exception) {
      Log.log("getfiles unable to access source folder, will retry");
    }
    return files;
  }
});


MMM-ImagesPhotos.js
/* global Log Module */

/* Magic Mirror
 * Module: MMM-ImagesPhotos
 *
 * By Rodrigo Ramírez Norambuena https://rodrigoramirez.com
 * MIT Licensed.
 */

Module.register("MMM-ImagesPhotos", {
  defaults: {
    opacity: 0.9,
    animationSpeed: 500,
    updateInterval: 5000,
    getInterval: 60000,
    maxWidth: "100%",
    maxHeight: "100%",
    retryDelay: 2500,
    path: "",
    fill: false,
    blur: 8
  },

  wrapper: null,
  suspended: false,
  timer: null,

  requiresVersion: "2.1.0", // Required version of MagicMirror

  start() {
    this.photos = [];
    this.loaded = false;
    this.lastPhotoIndex = -1;
    this.sendSocketNotification("CONFIG", this.config);
  },
  getStyles() {
    return ["MMM-ImagesPhotos.css"];
  },

  /*
   * getPhotos
   * Requests new data from api url helper
   *
   */
  getPhotos() {
    const urlApHelper = "/MMM-ImagesPhotos/photos";
    const self = this;
    let retry = true;

    const photosRequest = new XMLHttpRequest();
    photosRequest.open("GET", urlApHelper, true);
    photosRequest.onreadystatechange = function () {
      if (this.readyState === 4) {
        if (this.status === 200) {
          self.processPhotos(JSON.parse(this.response));
        } else if (this.status === 401) {
          self.updateDom(self.config.animationSpeed);
          Log.error(self.name, this.status);
          retry = false;
        } else {
          Log.error(self.name, "Could not load photos.");
        }

        if (retry) {
          self.scheduleUpdate(self.loaded ? -1 : self.config.retryDelay);
        }
      }
    };
    photosRequest.send();
  },
  notificationReceived() {
    // hook to turn off messages about notifications, clock once a second
  },
  startTimer() {
    const self = this;
    self.timer = setTimeout(() => {
      // clear timer value for resume
      self.timer = null;
      if (self.suspended === false) {
        self.updateDom(self.config.animationSpeed);
      }
    }, this.config.updateInterval);
  },

  socketNotificationReceived(notification) {
    if (notification === "READY") {
      // Schedule update timer.
      this.getPhotos();
    }
  },
  /* scheduleUpdate()
   * Schedule next update.
   *
   * argument delay number - Milliseconds before next update.
   *  If empty, this.config.updateInterval is used.
   */
  scheduleUpdate(delay) {
    let nextLoad = this.config.getInterval;
    if (typeof delay !== "undefined" && delay >= 0) {
      nextLoad = delay;
    }
    const self = this;
    setTimeout(function () {
      self.getPhotos();
    }, nextLoad);
  },

  /* randomIndex(photos)
   * Generate a random index for a list of photos.
   *
   * argument photos Array<String> - Array with photos.
   *
   * return Number - Random index.
   */
  randomIndex(photos) {
    if (photos.length === 1) {
      return 0;
    }

    const generate = function () {
      return Math.floor(Math.random() * photos.length);
    };

    const photoIndex = generate();
    this.lastPhotoIndex = photoIndex;

    return photoIndex;
  },

  /* complimentArray()
   * Retrieve a random photos.
   *
   * return photo Object - A photo.
   */
  randomPhoto() {
    const { photos } = this;
    const index = this.randomIndex(photos);

    return photos[index];
  },

  ScaleImage(srcwidth, srcheight, targetwidth, targetheight, fLetterBox) {
    const result = { width: 0, height: 0, fScaleToTargetWidth: true };

    if (
      srcwidth <= 0 ||
      srcheight <= 0 ||
      targetwidth <= 0 ||
      targetheight <= 0
    ) {
      return result;
    }

    // scale to the target width
    const scaleX1 = targetwidth;
    const scaleY1 = (srcheight * targetwidth) / srcwidth;

    // scale to the target height
    const scaleX2 = (srcwidth * targetheight) / srcheight;
    const scaleY2 = targetheight;

    // now figure out which one we should use
    let fScaleOnWidth = scaleX2 > targetwidth;
    if (fScaleOnWidth) {
      fScaleOnWidth = fLetterBox;
    } else {
      fScaleOnWidth = !fLetterBox;
    }

    if (fScaleOnWidth) {
      result.width = Math.floor(scaleX1);
      result.height = Math.floor(scaleY1);
      result.fScaleToTargetWidth = true;
    } else {
      result.width = Math.floor(scaleX2);
      result.height = Math.floor(scaleY2);
      result.fScaleToTargetWidth = false;
    }
    result.targetleft = Math.floor((targetwidth - result.width) / 2);
    result.targettop = Math.floor((targetheight - result.height) / 2);

    return result;
  },

  suspend() {
    this.suspended = true;
    if (this.timer != null) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  },
  resume() {
    this.suspended = false;
    if (this.timer == null) {
      this.startTimer();
    }
  },

  getDom() {
    const self = this;
    // if wrapper div not yet created
    if (this.wrapper == null) {
      // create it once, try to reduce image flash on change
      this.wrapper = document.createElement("div");
      this.bk = document.createElement("div");
      this.bk.className = "bgimage";
      if (this.config.fill === true) {
        this.bk.style.filter = `blur(${this.config.blur}px)`;
        this.bk.style["-webkit-filter"] = `blur(${this.config.blur}px)`;
      } else {
        this.bk.style.backgroundColor = this.config.backgroundColor;
      }
      this.wrapper.appendChild(this.bk);
      this.fg = document.createElement("div");
      this.wrapper.appendChild(this.fg);
    }
    if (this.photos.length) {
      // get the size of the margin, if any, we want to be full screen
      const m = window
        .getComputedStyle(document.body, null)
        .getPropertyValue("margin-top");
      // set the style for the containing div

      this.fg.style.border = "none";
      this.fg.style.margin = "0px";

      const photoImage = this.randomPhoto();
      let img = null;
      if (photoImage) {
        // create img tag element
        img = document.createElement("img");

        // set default position, corrected in onload handler
        img.style.left = `${0}px`;
        img.style.top = document.body.clientHeight + parseInt(m, 10) * 2;
        img.style.position = "relative";

        img.src = photoImage.url;
        // make invisible
        img.style.opacity = 0;
        // append this image to the div
        this.fg.appendChild(img);

        // set the onload event handler
        // the loadurl request will happen when the html is returned to MM and inserted into the dom.
        img.onload = (evt) => {
          // get the image of the event
          img = evt.currentTarget;
          Log.log(`image loaded=${img.src} size=${img.width}:${img.height}`);

          // what's the size of this image and it's parent
          const w = img.width;
          const h = img.height;
          const tw = document.body.clientWidth + parseInt(m, 10) * 2;
          const th = document.body.clientHeight + parseInt(m, 10) * 2;

          // compute the new size and offsets
          const result = self.ScaleImage(w, h, tw, th, true);

          // adjust the image size
          img.width = result.width;
          img.height = result.height;

          Log.log(`image setting size to ${result.width}:${result.height}`);
          Log.log(
            `image setting top to ${result.targetleft}:${result.targettop}`
          );

          // adjust the image position
          img.style.left = `${result.targetleft}px`;
          img.style.top = `${result.targettop}px`;

          // if another image was already displayed
          const c = self.fg.childElementCount;
          if (c > 1) {
            for (let i = 0; i < c - 1; i += 1) {
              // hide it
              self.fg.firstChild.style.opacity = 0;
              self.fg.firstChild.style.backgroundColor = "rgba(0,0,0,0)";
              // remove the image element from the div
              self.fg.removeChild(self.fg.firstChild);
            }
          }
          self.fg.firstChild.style.opacity = self.config.opacity;

          self.fg.firstChild.style.transition = "opacity 1.25s";
          if (self.config.fill === true) {
            self.bk.style.backgroundImage = `url(${self.fg.firstChild.src})`;
          }
          self.startTimer();
        };
      }
    }
    return this.wrapper;
  },

  getScripts() {
    return ["MMM-ImagesPhotos.css"];
  },

  processPhotos(data) {
    const self = this;
    this.photos = data;
    if (this.loaded === false) {
      if (this.suspended === false) {
        self.updateDom(self.config.animationSpeed);
      }
    }
    this.loaded = true;
  }
});
